### PR TITLE
Add essential verification logic

### PR DESCRIPTION
**Pull request by junha1**

### CODE REVIEW

File: common/src/lib.rs [hunk 1]
```diff
@@ -27,4 +29,77 @@ impl BlockHeader {
...
+            ));
+        }
+        if header.previous_hash != self.hash() {
+            return Err(format!(
+                "Invalid previous hash: expected {}, got {}",
+                self.hash(),
+                header.previous_hash
+            ));
+        }
+        if !self
```
Comment (byeongjee) [inline common/src/lib.rs:51]: Comparison with the current leader should be done here, right?
Then would you add a TODO comment for that?

File: common/src/lib.rs [hunk 2]
```diff
@@ -27,4 +29,77 @@ impl BlockHeader {
...
+        &self,
+        block_finalization_proof: &[(PublicKey, Signature)],
+    ) -> Result<(), String> {
+        let total_voting_power: u64 = self.validator_set.iter().map(|(_, v)| v).sum();
+        // TODO: change to `HashSet` after `PublicKey` supports `Hash`.
+        let mut voted_validators = BTreeSet::new();
+        for (public_key, signature) in block_finalization_proof {
+            if !signature.verify(self.hash(), public_key) {
+                return Err(format!(
+                    "Invalid finalization proof: {}, {}",
```
Comment (byeongjee) [inline common/src/lib.rs:85]: How about `Invalid finailzation proof: signature verification failure : ..`?

File: common/src/lib.rs [hunk 3]
```diff
@@ -27,4 +29,77 @@ impl BlockHeader {
...
+                    public_key, signature
+                ));
+            }
+            voted_validators.insert(public_key);
+        }
+        let mut voted_voting_power: u64 = 0;
+        for (validator, votin_power) in &self.validator_set {
```
Comment (byeongjee) [inline common/src/lib.rs:92]: How about `self.validator_set.inter().filter(...).sum(...)`?

Comment (byeongjee) [inline common/src/lib.rs:92]: typo: votin_power -> voting_power

File: common/src/lib.rs [hunk 4]
```diff
@@ -27,4 +29,77 @@ impl BlockHeader {
...
+            ));
+        }
+        if !self
+            .validator_set
+            .iter()
+            .any(|(pk, _)| pk == &header.author)
+        {
+            return Err(format!("Invalid author: got {}", header.author));
+        }
+        if header.timestamp < self.timestamp {
```
Comment (TomTaehoonKim) [inline common/src/lib.rs:58]: What if in equal case? Does it matter?

File: common/src/lib.rs (hunk 3 above)
Comment (TomTaehoonKim) [reply common/src/lib.rs:92]: I personally think Junha's code is better in terms of readability.

File: common/src/lib.rs [hunk 5]
```diff
@@ -27,4 +29,77 @@ impl BlockHeader {
...
+                    "Invalid finalization proof: {}, {}",
+                    public_key, signature
+                ));
+            }
+            voted_validators.insert(public_key);
+        }
+        let mut voted_voting_power: u64 = 0;
+        for (validator, votin_power) in &self.validator_set {
+            if voted_validators.contains(validator) {
+                voted_voting_power += votin_power;
```
Comment (TomTaehoonKim) [inline common/src/lib.rs:94]: same typo here: votin_power -> voting_power

File: common/src/lib.rs (hunk 4 above)
Comment (junha1) [reply common/src/lib.rs:60]: should be rejected.

File: common/src/lib.rs (hunk 2 above)
Comment (junha1) [reply common/src/lib.rs]: fixed

File: common/src/lib.rs (hunk 3 above)
Comment (junha1) [reply common/src/lib.rs]: fixed

Comment (junha1) [reply common/src/lib.rs]: fixed

File: common/src/lib.rs (hunk 5 above)
Comment (junha1) [reply common/src/lib.rs]: fixed

**Review by yeomjh00 — APPROVED** [review]

**Pull request merged**